docs/limitrange.md

```diff
@@ -0,0 +1,104 @@
+
+A _LimitRange_ provides constraints that can:
+
+- Enforce minimum and maximum compute resources usage per Pod or Container in a namespace.
+- Enforce minimum and maximum storage request per PersistentVolumeClaim in a namespace.
+- Enforce a ratio between request and limit for a resource in a namespace.
+- Set default request/limit for compute resources in a namespace and automatically inject them to Containers at runtime.
+
+`LimitRange` are validating and mutating `Requests` and `Limits`. Let's look, *in a nutshell*, on how those work in Kubernetes.
+
+- **Requests** are not enforced. If the node has more resource available than the request, the container can use it.
+- **Limits** on the other hand, are a hard stop. A container going over the limit, will be killed.
+
+Resource types for both are:
+-   CPU
+-   Memory
+-   Ephemeral storage
+
+The next question is : how pods with resource and limits are run/scheduled ?
+The scheduler *computes* the amount of CPA and memory requests (using **Requests**) and tries to find a node to schedule it.
+
+Requests and limits can be applied to both Containers and Init Containers.
+- For init containers, the max of each type is taken
+- For containers, it sums all requests/limits for each containers
+
+This means, if you got the following:
+-   initContainer1 : 1 CPU, 100m memory
+-   initContainer2 : 2 CPU, 200m memory
+-   container1 : 1 CPU, 50m memory
+-   container2 : 2 CPU, 250m memory
+-   container3 : 3 CPU, 500m memory
+
+The computation will be:
+-   CPU : 2 (max init containers) + 6 (sum of containers) = 8 CPU
+-   Memory: 200m (max init containers) + 800m (sum of containers) = 1000m (1G)
+
+## Tekton support
+
+The way Limits and Requests works in Kubernetes is because it is assumed that all containers run in parallel, and init container run before, each one after the others.
+
+That assumption — containers running in parallel — is not true in Tekton. They do all start together (because there is no way around this) **but** the *entrypoint hack* is making sure they actually run in sequence and thus there is always only one container that is actually consuming some resource at the same time.
+
+This means, we need to handle limits, request and LimitRanges in a *non-standard* way. Let's try to define that. Tekton needs to take into account all the aspect of the LimitRange : the min/max as well as the default. If there is no default, but there is min/max, Tekton need then to **set** a default value that is between the min/max. If we set the value too low, the Pod won't be able to be created, similar if we set the value too high. **But** those values are set on **containers**, so we **have to** do our own computation to know what request to put on each containers.
```

Comment:
I got a bit confused by this part - specifically why we need to set the defaults.

After reading through everything I think the overall reason is that we can't rely on the existing LimitRange mutating webhook, so we pretty much need to explicitly set everything ourselves, is that right?

Maybe you could emphasize that a bit here, something like:

```
This means, we need to handle limits, request and LimitRanges in a *non-standard* way. Since the existing LimitRange mutating webhook won't take Tekton's requirements into account, Tekton needs to fully control all the values the LimitRange webhook might set...
```

Not sure if that's clearer, feel free to ignore 🙏

Reply:
> After reading through everything I think the overall reason is that we can't rely on the existing LimitRange mutating webhook, so we pretty much need to explicitly set everything ourselves, is that right?

Yes indeed. If we want allocate resource in *relatively* optimial manner, we cannot rely on the LimitRange mutating webhook.